{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "The recent development of Hi-C and related methods have revealed substantial hierarchical structures of 3D chromatin organization, such as compartments, sub-compartments, topologically associating domains, and loops. However, we still know little about the heterogeneity and variability of these structures in single cells as the measurements derived from Hi-C are based on the population of millions of cells. For example, to what extent the Hi-C interaction frequency related to the physical distance for two genomic loci. To answer these kinds of questions,  Finn et al. applied HiFISH, a high-throughput imaging method, to assess the relationship between the spatial location and colocalization frequencies for more than a hundred pairs of genomic loci in single-cell resolution. \n",
    "\n",
    "In this notebook, we aim to use the Nucleome Browser to interactively explore the processed data from Finn et al. Specifically, we will use the Python notebook to examine the correlation between Hi-C interaction frequency and contact frequency of pairs of genomic loci. Nucleome Browser has a specific panel hosting microscopy data from the DCIC data portal. We will use this panel to explore raw FISH images on Nucleome Browser and view them interactively inside the notebook. We will also use a customized web page to view the Hi-C matrix on the HiGlass platform. Notebook, FISH images on Nucleome Browser, and Hi-C contact matrix in HiGlass are all connected through the unique 'event-driven communication' developed in the Nucleome Browser."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Microscopy Data Preprocessing\n",
    "In this section, we will retrieve the Hi-C data and processed image data of Finn et al. We will use the 4DN DCIC Jupyter notebook to directly access the data. THe script in this sections requires access to the 4DN DCIC Jupyter Hub. You also prepare the processed data if you cannot run this notebook in the 4DN DCIC Jupyter notebook server. You can skip this session and load the processed data directly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dcicutils import jh_utils\n",
    "import numpy\n",
    "import pandas as pd\n",
    "import cooler\n",
    "\n",
    "# the mcool file from the dataset\n",
    "file_path_hic = jh_utils.mount_4dn_file('4DNFIMDOXUT8')\n",
    "\n",
    "# the summary results file for microscopy images\n",
    "file_path_mic = jh_utils.mount_4dn_file('4DNFID78GRES')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then load the processed file for the microscopy data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bait ID</th>\n",
       "      <th>Bait Chr</th>\n",
       "      <th>Bait Start</th>\n",
       "      <th>Bait End</th>\n",
       "      <th>Target ID</th>\n",
       "      <th>Target Chr</th>\n",
       "      <th>Target Start</th>\n",
       "      <th>Target End</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Hi-C Capture Frequency</th>\n",
       "      <th>...</th>\n",
       "      <th>Fig 1 (E-I)</th>\n",
       "      <th>Fig 1 (J-L)</th>\n",
       "      <th>Fig 2</th>\n",
       "      <th>Fig 3</th>\n",
       "      <th>Fig 4</th>\n",
       "      <th>Fig 5</th>\n",
       "      <th>Fig 6</th>\n",
       "      <th>Bait</th>\n",
       "      <th>Target</th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>misteli:probe_11</td>\n",
       "      <td>1</td>\n",
       "      <td>2370445</td>\n",
       "      <td>2570720</td>\n",
       "      <td>misteli:probe_74</td>\n",
       "      <td>1</td>\n",
       "      <td>18566409</td>\n",
       "      <td>18766880</td>\n",
       "      <td>Chr1GR</td>\n",
       "      <td>0.001959</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>All</td>\n",
       "      <td>D,E</td>\n",
       "      <td>All</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>misteli:probe_11</td>\n",
       "      <td>1</td>\n",
       "      <td>2370445</td>\n",
       "      <td>2570720</td>\n",
       "      <td>misteli:probe_91</td>\n",
       "      <td>1</td>\n",
       "      <td>22223362</td>\n",
       "      <td>22394657</td>\n",
       "      <td>Chr1GR</td>\n",
       "      <td>0.000818</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>All</td>\n",
       "      <td>D,E</td>\n",
       "      <td>All</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>misteli:probe_11</td>\n",
       "      <td>1</td>\n",
       "      <td>2370445</td>\n",
       "      <td>2570720</td>\n",
       "      <td>misteli:probe_441</td>\n",
       "      <td>1</td>\n",
       "      <td>109469002</td>\n",
       "      <td>109662490</td>\n",
       "      <td>Chr1GR</td>\n",
       "      <td>0.000194</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>All</td>\n",
       "      <td>D,E</td>\n",
       "      <td>All</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "      <td>441.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Bait ID  Bait Chr  Bait Start  Bait End          Target ID  \\\n",
       "0  misteli:probe_11         1     2370445   2570720   misteli:probe_74   \n",
       "1  misteli:probe_11         1     2370445   2570720   misteli:probe_91   \n",
       "2  misteli:probe_11         1     2370445   2570720  misteli:probe_441   \n",
       "\n",
       "   Target Chr  Target Start  Target End Dataset  Hi-C Capture Frequency  ...  \\\n",
       "0           1      18566409    18766880  Chr1GR                0.001959  ...   \n",
       "1           1      22223362    22394657  Chr1GR                0.000818  ...   \n",
       "2           1     109469002   109662490  Chr1GR                0.000194  ...   \n",
       "\n",
       "   Fig 1 (E-I) Fig 1 (J-L) Fig 2  Fig 3  Fig 4 Fig 5 Fig 6  Bait  Target  chr  \n",
       "0          NaN         All   D,E    All    NaN   NaN   NaN  11.0    74.0    1  \n",
       "1          NaN         All   D,E    All    NaN   NaN   NaN  11.0    91.0    1  \n",
       "2          NaN         All   D,E    All    NaN   NaN   NaN  11.0   441.0    1  \n",
       "\n",
       "[3 rows x 33 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the processed data\n",
    "mic_data = pd.read_csv(file_path_mic, sep = '\\t', header = 0)\n",
    "mic_data.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Bait ID', 'Bait Chr', 'Bait Start', 'Bait End', 'Target ID',\n",
       "       'Target Chr', 'Target Start', 'Target End', 'Dataset',\n",
       "       'Hi-C Capture Frequency', 'O/E Hi-C', 'Category', 'Gene Content',\n",
       "       'Bias Quantile', 'Genomic Distance', 'Bait Compartment',\n",
       "       'Target Compartment', 'nspots', '% within 150 nm (3D)',\n",
       "       '% within 200 nm (3D)', '% within 350 nm (3D)', '% within 1 µm (3D)',\n",
       "       'Mean 3D distance', 'Fig 1 (E-I)', 'Fig 1 (J-L)', 'Fig 2', 'Fig 3',\n",
       "       'Fig 4', 'Fig 5', 'Fig 6', 'Bait', 'Target', 'chr'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check column names\n",
    "mic_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(128, 33)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check how many pairs of loci\n",
    "mic_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are in total 128 pairs of loci. We then add unique IDs (strings) for each bait and target by concatenating chromosome name, start, and ending position into the format 'chr:start-end' following the style of the UCSC Genome Browser."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "mic_data['Bait_string'] = 'chr' + mic_data['Bait Chr'].astype(str) + ':' + mic_data['Bait Start'].astype(str) + '-' + mic_data['Bait End'].astype(str)\n",
    "mic_data['Target_string'] = 'chr' + mic_data['Target Chr'].astype(str) + ':' + mic_data['Target Start'].astype(str) + '-' + mic_data['Target End'].astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the string of Bait and Target."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bait ID</th>\n",
       "      <th>Bait_string</th>\n",
       "      <th>Target ID</th>\n",
       "      <th>Target_string</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>misteli:probe_11</td>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>misteli:probe_74</td>\n",
       "      <td>chr1:18566409-18766880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>misteli:probe_11</td>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>misteli:probe_91</td>\n",
       "      <td>chr1:22223362-22394657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>misteli:probe_11</td>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>misteli:probe_441</td>\n",
       "      <td>chr1:109469002-109662490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>misteli:probe_11</td>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>misteli:probe_994</td>\n",
       "      <td>chr1:248125293-248263553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>misteli:probe_52</td>\n",
       "      <td>chr1:12708719-12865743</td>\n",
       "      <td>misteli:probe_11</td>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>misteli:probe_52</td>\n",
       "      <td>chr1:12708719-12865743</td>\n",
       "      <td>misteli:probe_74</td>\n",
       "      <td>chr1:18566409-18766880</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Bait ID             Bait_string          Target ID  \\\n",
       "0  misteli:probe_11    chr1:2370445-2570720   misteli:probe_74   \n",
       "1  misteli:probe_11    chr1:2370445-2570720   misteli:probe_91   \n",
       "2  misteli:probe_11    chr1:2370445-2570720  misteli:probe_441   \n",
       "3  misteli:probe_11    chr1:2370445-2570720  misteli:probe_994   \n",
       "4  misteli:probe_52  chr1:12708719-12865743   misteli:probe_11   \n",
       "5  misteli:probe_52  chr1:12708719-12865743   misteli:probe_74   \n",
       "\n",
       "              Target_string  \n",
       "0    chr1:18566409-18766880  \n",
       "1    chr1:22223362-22394657  \n",
       "2  chr1:109469002-109662490  \n",
       "3  chr1:248125293-248263553  \n",
       "4      chr1:2370445-2570720  \n",
       "5    chr1:18566409-18766880  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mic_data.loc[0:5, ['Bait ID', 'Bait_string', 'Target ID', 'Target_string']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate Hi-C interaction intensity\n",
    "In this section, we will get the corresponding Hi-C interaction intensity for these 128 pairs of loci."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each bait and target, we first identify the closest bin ID that covers the region between the bait and the target. We then retrieve the distance normalized contact frequencies. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_hic_intensity(mcoolfile, mic_data, resolution, valuetype):\n",
    "    hic = cooler.Cooler(mcoolfile + '::resolutions/' + str(resolution))\n",
    "    result = []\n",
    "    for index, row in mic_data.iterrows():\n",
    "        start_1 = list(hic.bins().fetch(row['Bait_string']).index)[0]\n",
    "        end_1 = list(hic.bins().fetch(row['Bait_string']).index)[-1] + 1\n",
    "        start_2 = list(hic.bins().fetch(row['Target_string']).index)[0]\n",
    "        end_2 = list(hic.bins().fetch(row['Target_string']).index)[-1] + 1\n",
    "        # normalized frequencies\n",
    "        values = hic.matrix(balance=True, as_pixels=True, join=True)[start_1:end_1, start_2:end_2][valuetype]\n",
    "        # average per-bin\n",
    "        nbins1 = end_1 - start_1\n",
    "        nbins2 = end_2 - start_2\n",
    "        s = numpy.nansum(values)   # sum of per-bin intensities\n",
    "        m = s / (nbins1 * nbins2)   # per-bin intensity\n",
    "        m = m / resolution   # control for resolution effect\n",
    "        result.append(m)\n",
    "    return result\n",
    "# get the normalized Hi-C interaction intensity\n",
    "hic_intensity = get_hic_intensity(file_path_hic, mic_data, 10000, valuetype = 'balanced')\n",
    "mic_data['hic_10k'] = hic_intensity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Bait_string</th>\n",
       "      <th>Target_string</th>\n",
       "      <th>hic_interaction</th>\n",
       "      <th>fraction_ls_350</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>chr1:18566409-18766880</td>\n",
       "      <td>3.692195e-09</td>\n",
       "      <td>2.4120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>chr1:22223362-22394657</td>\n",
       "      <td>1.133845e-09</td>\n",
       "      <td>0.8362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>chr1:109469002-109662490</td>\n",
       "      <td>5.075978e-10</td>\n",
       "      <td>1.0030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>chr1:248125293-248263553</td>\n",
       "      <td>5.806794e-10</td>\n",
       "      <td>1.4553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:12708719-12865743</td>\n",
       "      <td>chr1:2370445-2570720</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>5.0270</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Bait_string             Target_string  hic_interaction  \\\n",
       "0    chr1:2370445-2570720    chr1:18566409-18766880     3.692195e-09   \n",
       "1    chr1:2370445-2570720    chr1:22223362-22394657     1.133845e-09   \n",
       "2    chr1:2370445-2570720  chr1:109469002-109662490     5.075978e-10   \n",
       "3    chr1:2370445-2570720  chr1:248125293-248263553     5.806794e-10   \n",
       "4  chr1:12708719-12865743      chr1:2370445-2570720     0.000000e+00   \n",
       "\n",
       "   fraction_ls_350  \n",
       "0           2.4120  \n",
       "1           0.8362  \n",
       "2           1.0030  \n",
       "3           1.4553  \n",
       "4           5.0270  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = mic_data[['Bait_string', 'Target_string', 'hic_10k', '% within 350 nm (3D)']]\n",
    "data.columns = ['Bait_string', 'Target_string', 'hic_interaction', 'fraction_ls_350']\n",
    "# save the processed data into a file\n",
    "data.to_csv(\"Finn_et_all_processed_data.tsv\", sep = '\\t', header = True, index = False)\n",
    "# convert \n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create an interactive scatterplot by linking to the Nucleome Browser\n",
    "In this section, we will plot an interactive scatterplot with X-axis as the fraction of genomic loci within 350nm in spatial distance and Y-axis as the calculated Hi-C interaction intensity. Users can then click the points and view the Nucleome Browser will navigate to the corresponding regions and highlight the bait and target probe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can directly load the processed data if you do not have the access to the 4DN DCIC Jupyter server.\n",
    "\n",
    "> **Nucleome Browser (version later than v0.9.9) supports the users to set channel ID for communication. In this demo, we will use the default communication channel (N-cnbChan01). If you want to use other channel (e.g., N-cnbChan02). you need to switch the channel using dispatch.chanId(\"<new channel>\").**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "data = pd.read_csv(\"data/Finn_et_all_processed_data.tsv\", sep = '\\t', header = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display, Javascript, HTML\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "require.config({\n",
       "    paths: {\n",
       "        'd3': ['https://d3js.org/d3.v5.min'],\n",
       "        'nb': ['https://vis.nucleome.org/static/lib/nb-dispatch'],\n",
       "        'Plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "    }\n",
       "});\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%javascript\n",
    "require.config({\n",
    "    paths: {\n",
    "        'd3': ['https://d3js.org/d3.v5.min'],\n",
    "        'nb': ['https://vis.nucleome.org/static/lib/nb-dispatch'],\n",
    "        'Plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
    "    }\n",
    "});"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(element){\n",
       "    require(['d3', 'nb', 'Plotly'], function(d3, nb, Plotly) {\n",
       "        // init div for plot\n",
       "        d3.select(element[0]).append('div')\n",
       "            .attr('id', 'myDiv')\n",
       "        var myPlot = document.getElementById('myDiv');\n",
       "        // prepare data\n",
       "        var df = {\"Bait_string\":{\"0\":\"chr1:2370445-2570720\",\"1\":\"chr1:2370445-2570720\",\"2\":\"chr1:2370445-2570720\",\"3\":\"chr1:2370445-2570720\",\"4\":\"chr1:12708719-12865743\",\"5\":\"chr1:12708719-12865743\",\"6\":\"chr1:12708719-12865743\",\"7\":\"chr1:12708719-12865743\",\"8\":\"chr1:12708719-12865743\",\"9\":\"chr1:22223362-22394657\",\"10\":\"chr1:96067992-96226766\",\"11\":\"chr1:96067992-96226766\",\"12\":\"chr1:96067992-96226766\",\"13\":\"chr1:97025755-97207860\",\"14\":\"chr1:105256970-105427280\",\"15\":\"chr1:105256970-105427280\",\"16\":\"chr1:105256970-105427280\",\"17\":\"chr1:105256970-105427280\",\"18\":\"chr1:105256970-105427280\",\"19\":\"chr1:105256970-105427280\",\"20\":\"chr1:109469002-109662490\",\"21\":\"chr1:115549997-115729561\",\"22\":\"chr1:115549997-115729561\",\"23\":\"chr1:115549997-115729561\",\"24\":\"chr1:115549997-115729561\",\"25\":\"chr1:118815746-118989848\",\"26\":\"chr1:118815746-118989848\",\"27\":\"chr1:118815746-118989848\",\"28\":\"chr1:118815746-118989848\",\"29\":\"chr1:118815746-118989848\",\"30\":\"chr4:54270488-54435016\",\"31\":\"chr4:54270488-54435016\",\"32\":\"chr4:54270488-54435016\",\"33\":\"chr4:54270488-54435016\",\"34\":\"chr4:54270488-54435016\",\"35\":\"chr4:54270488-54435016\",\"36\":\"chr4:54270488-54435016\",\"37\":\"chr4:54542457-54695030\",\"38\":\"chr4:54542457-54695030\",\"39\":\"chr4:54542457-54695030\",\"40\":\"chr4:54542457-54695030\",\"41\":\"chr4:54542457-54695030\",\"42\":\"chr4:54542457-54695030\",\"43\":\"chr4:54542457-54695030\",\"44\":\"chr4:54794388-54995522\",\"45\":\"chr4:54794388-54995522\",\"46\":\"chr4:54794388-54995522\",\"47\":\"chr4:54794388-54995522\",\"48\":\"chr4:54794388-54995522\",\"49\":\"chr4:54794388-54995522\",\"50\":\"chr4:55053142-55196692\",\"51\":\"chr4:55053142-55196692\",\"52\":\"chr4:55236649-55437873\",\"53\":\"chr4:55586513-55775729\",\"54\":\"chr4:55586513-55775729\",\"55\":\"chr4:56020676-56178582\",\"56\":\"chr4:150978277-151124968\",\"57\":\"chr4:150978277-151124968\",\"58\":\"chr4:150978277-151124968\",\"59\":\"chr4:150978277-151124968\",\"60\":\"chr4:150978277-151124968\",\"61\":\"chr4:150978277-151124968\",\"62\":\"chr4:150978277-151124968\",\"63\":\"chr4:150978277-151124968\",\"64\":\"chr4:151276234-151444375\",\"65\":\"chr4:151276234-151444375\",\"66\":\"chr4:151276234-151444375\",\"67\":\"chr4:151276234-151444375\",\"68\":\"chr4:151276234-151444375\",\"69\":\"chr4:151276234-151444375\",\"70\":\"chr4:151542617-151679223\",\"71\":\"chr4:151542617-151679223\",\"72\":\"chr4:151542617-151679223\",\"73\":\"chr4:151542617-151679223\",\"74\":\"chr4:151542617-151679223\",\"75\":\"chr4:151542617-151679223\",\"76\":\"chr4:151942840-152102227\",\"77\":\"chr4:151942840-152102227\",\"78\":\"chr4:151942840-152102227\",\"79\":\"chr4:151942840-152102227\",\"80\":\"chr4:151942840-152102227\",\"81\":\"chr4:152521628-152677396\",\"82\":\"chr4:152521628-152677396\",\"83\":\"chr4:152712784-152899106\",\"84\":\"chr4:152897610-153104002\",\"85\":\"chr4:152897610-153104002\",\"86\":\"chr4:152897610-153104002\",\"87\":\"chr4:153155365-153328553\",\"88\":\"chr4:153347384-153505952\",\"89\":\"chr17:30088606-30263568\",\"90\":\"chr17:30088606-30263568\",\"91\":\"chr17:30088606-30263568\",\"92\":\"chr17:30088606-30263568\",\"93\":\"chr17:30088606-30263568\",\"94\":\"chr17:30088606-30263568\",\"95\":\"chr17:35268046-35441605\",\"96\":\"chr17:35268046-35441605\",\"97\":\"chr17:35268046-35441605\",\"98\":\"chr17:35268046-35441605\",\"99\":\"chr17:35268046-35441605\",\"100\":\"chr17:36387556-36574537\",\"101\":\"chr17:36387556-36574537\",\"102\":\"chr17:36387556-36574537\",\"103\":\"chr17:36387556-36574537\",\"104\":\"chr17:36387556-36574537\",\"105\":\"chr17:38743489-38939529\",\"106\":\"chr17:38743489-38939529\",\"107\":\"chr17:38743489-38939529\",\"108\":\"chr17:48502126-48689497\",\"109\":\"chr17:48502126-48689497\",\"110\":\"chr17:48502126-48689497\",\"111\":\"chr17:48502126-48689497\",\"112\":\"chr17:65509811-65678200\",\"113\":\"chr18:2535851-2706101\",\"114\":\"chr18:2535851-2706101\",\"115\":\"chr18:2535851-2706101\",\"116\":\"chr18:2535851-2706101\",\"117\":\"chr18:2535851-2706101\",\"118\":\"chr18:22075002-22239510\",\"119\":\"chr18:34992711-35151964\",\"120\":\"chr18:34992711-35151964\",\"121\":\"chr18:62276743-62456847\",\"122\":\"chr18:62276743-62456847\",\"123\":\"chr18:62276743-62456847\",\"124\":\"chr18:75525179-75695249\",\"125\":\"chr18:75525179-75695249\",\"126\":\"chr18:75525179-75695249\",\"127\":\"chr18:75525179-75695249\"},\"Target_string\":{\"0\":\"chr1:18566409-18766880\",\"1\":\"chr1:22223362-22394657\",\"2\":\"chr1:109469002-109662490\",\"3\":\"chr1:248125293-248263553\",\"4\":\"chr1:2370445-2570720\",\"5\":\"chr1:18566409-18766880\",\"6\":\"chr1:22223362-22394657\",\"7\":\"chr1:109469002-109662490\",\"8\":\"chr1:248125293-248263553\",\"9\":\"chr1:18566409-18766880\",\"10\":\"chr1:88222722-88418471\",\"11\":\"chr1:89759075-89913486\",\"12\":\"chr1:97025755-97207860\",\"13\":\"chr1:89759075-89913486\",\"14\":\"chr1:88222722-88418471\",\"15\":\"chr1:89759075-89913486\",\"16\":\"chr1:96067992-96226766\",\"17\":\"chr1:97025755-97207860\",\"18\":\"chr1:115549997-115729561\",\"19\":\"chr1:118815746-118989848\",\"20\":\"chr1:248125293-248263553\",\"21\":\"chr1:88222722-88418471\",\"22\":\"chr1:89759075-89913486\",\"23\":\"chr1:96067992-96226766\",\"24\":\"chr1:97025755-97207860\",\"25\":\"chr1:88222722-88418471\",\"26\":\"chr1:89759075-89913486\",\"27\":\"chr1:96067992-96226766\",\"28\":\"chr1:97025755-97207860\",\"29\":\"chr1:115549997-115729561\",\"30\":\"chr4:54542457-54695030\",\"31\":\"chr4:54794388-54995522\",\"32\":\"chr4:55236649-55437873\",\"33\":\"chr4:55367316-55516239\",\"34\":\"chr4:55586513-55775729\",\"35\":\"chr4:56020676-56178582\",\"36\":\"chr4:56986598-57145686\",\"37\":\"chr4:54794388-54995522\",\"38\":\"chr4:55053142-55196692\",\"39\":\"chr4:55236649-55437873\",\"40\":\"chr4:55367316-55516239\",\"41\":\"chr4:55586513-55775729\",\"42\":\"chr4:56020676-56178582\",\"43\":\"chr4:56986598-57145686\",\"44\":\"chr4:55053142-55196692\",\"45\":\"chr4:55236649-55437873\",\"46\":\"chr4:55367316-55516239\",\"47\":\"chr4:55586513-55775729\",\"48\":\"chr4:56020676-56178582\",\"49\":\"chr4:56986598-57145686\",\"50\":\"chr4:55367316-55516239\",\"51\":\"chr4:55439495-55586511\",\"52\":\"chr4:55367316-55516239\",\"53\":\"chr4:56020676-56178582\",\"54\":\"chr4:56986598-57145686\",\"55\":\"chr4:56986598-57145686\",\"56\":\"chr4:151276234-151444375\",\"57\":\"chr4:151542617-151679223\",\"58\":\"chr4:151942840-152102227\",\"59\":\"chr4:152521628-152677396\",\"60\":\"chr4:152897610-153104002\",\"61\":\"chr4:153155365-153328553\",\"62\":\"chr4:153347384-153505952\",\"63\":\"chr4:153567330-153719454\",\"64\":\"chr4:151942840-152102227\",\"65\":\"chr4:152521628-152677396\",\"66\":\"chr4:152897610-153104002\",\"67\":\"chr4:153155365-153328553\",\"68\":\"chr4:153347384-153505952\",\"69\":\"chr4:153567330-153719454\",\"70\":\"chr4:151942840-152102227\",\"71\":\"chr4:152521628-152677396\",\"72\":\"chr4:152897610-153104002\",\"73\":\"chr4:153155365-153328553\",\"74\":\"chr4:153347384-153505952\",\"75\":\"chr4:153567330-153719454\",\"76\":\"chr4:152521628-152677396\",\"77\":\"chr4:152897610-153104002\",\"78\":\"chr4:153155365-153328553\",\"79\":\"chr4:153347384-153505952\",\"80\":\"chr4:153567330-153719454\",\"81\":\"chr4:152897610-153104002\",\"82\":\"chr4:153567330-153719454\",\"83\":\"chr4:153155365-153328553\",\"84\":\"chr4:153155365-153328553\",\"85\":\"chr4:153347384-153505952\",\"86\":\"chr4:153567330-153719454\",\"87\":\"chr4:153567330-153719454\",\"88\":\"chr4:153567330-153719454\",\"89\":\"chr17:9734635-9900986\",\"90\":\"chr17:35268046-35441605\",\"91\":\"chr17:36387556-36574537\",\"92\":\"chr17:38743489-38939529\",\"93\":\"chr17:48502126-48689497\",\"94\":\"chr17:65509811-65678200\",\"95\":\"chr17:9734635-9900986\",\"96\":\"chr17:38743489-38939529\",\"97\":\"chr17:48502126-48689497\",\"98\":\"chr17:51274882-51467771\",\"99\":\"chr17:65509811-65678200\",\"100\":\"chr17:9734635-9900986\",\"101\":\"chr17:35268046-35441605\",\"102\":\"chr17:38743489-38939529\",\"103\":\"chr17:51274882-51467771\",\"104\":\"chr17:65509811-65678200\",\"105\":\"chr17:9734635-9900986\",\"106\":\"chr17:48502126-48689497\",\"107\":\"chr17:51274882-51467771\",\"108\":\"chr17:9734635-9900986\",\"109\":\"chr17:36387556-36574537\",\"110\":\"chr17:51274882-51467771\",\"111\":\"chr17:65509811-65678200\",\"112\":\"chr17:38743489-38939529\",\"113\":\"chr18:22075002-22239510\",\"114\":\"chr18:34992711-35151964\",\"115\":\"chr18:46316193-46486760\",\"116\":\"chr18:62276743-62456847\",\"117\":\"chr18:75525179-75695249\",\"118\":\"chr18:46316193-46486760\",\"119\":\"chr18:22075002-22239510\",\"120\":\"chr18:46316193-46486760\",\"121\":\"chr18:22075002-22239510\",\"122\":\"chr18:34992711-35151964\",\"123\":\"chr18:46316193-46486760\",\"124\":\"chr18:22075002-22239510\",\"125\":\"chr18:34992711-35151964\",\"126\":\"chr18:46316193-46486760\",\"127\":\"chr18:62276743-62456847\"},\"hic_interaction\":{\"0\":0.0000000037,\"1\":0.0000000011,\"2\":0.0000000005,\"3\":0.0000000006,\"4\":0.0,\"5\":0.000000011,\"6\":0.0000000009,\"7\":0.0000000003,\"8\":0.000000002,\"9\":0.0,\"10\":0.0,\"11\":0.0,\"12\":0.0000000236,\"13\":0.0,\"14\":0.0,\"15\":0.0,\"16\":0.0,\"17\":0.0,\"18\":0.0000000035,\"19\":0.0000000053,\"20\":0.0,\"21\":0.0,\"22\":0.0,\"23\":0.0,\"24\":0.0,\"25\":0.0,\"26\":0.0,\"27\":0.0,\"28\":0.0,\"29\":0.0,\"30\":0.0000001201,\"31\":0.0000001098,\"32\":0.0000000598,\"33\":0.0000000426,\"34\":0.0000000312,\"35\":0.0000000227,\"36\":0.0000000118,\"37\":0.0000001171,\"38\":0.0000000543,\"39\":0.0000000533,\"40\":0.0000000457,\"41\":0.0000000344,\"42\":0.0000000202,\"43\":0.0000000128,\"44\":0.0000001,\"45\":0.0000000868,\"46\":0.0000000458,\"47\":0.0000000586,\"48\":0.0000000344,\"49\":0.0000000092,\"50\":0.0000000606,\"51\":0.0000000633,\"52\":0.0000006368,\"53\":0.0000000773,\"54\":0.0000000113,\"55\":0.0000000117,\"56\":0.000000088,\"57\":0.0000000615,\"58\":0.0000000395,\"59\":0.0000000248,\"60\":0.0000000224,\"61\":0.0000000156,\"62\":0.0000000129,\"63\":0.0000000079,\"64\":0.000000076,\"65\":0.0000000355,\"66\":0.0000000222,\"67\":0.0000000182,\"68\":0.0000000155,\"69\":0.0000000075,\"70\":0.0000000796,\"71\":0.0000000363,\"72\":0.0000000165,\"73\":0.0000000121,\"74\":0.0000000138,\"75\":0.0000000078,\"76\":0.0000000518,\"77\":0.0000000253,\"78\":0.000000018,\"79\":0.0000000151,\"80\":0.0000000085,\"81\":0.0000000664,\"82\":0.0000000184,\"83\":0.0000000645,\"84\":0.0000000916,\"85\":0.0000000722,\"86\":0.0000000548,\"87\":0.0000001035,\"88\":0.0000001423,\"89\":0.0,\"90\":0.0000000034,\"91\":0.0000000028,\"92\":0.0000000055,\"93\":0.000000003,\"94\":0.0000000015,\"95\":0.0,\"96\":0.000000005,\"97\":0.000000002,\"98\":0.000000004,\"99\":0.0000000017,\"100\":0.0,\"101\":0.0,\"102\":0.0000000045,\"103\":0.0000000036,\"104\":0.0000000008,\"105\":0.0,\"106\":0.0000000042,\"107\":0.000000001,\"108\":0.0,\"109\":0.0,\"110\":0.0000000026,\"111\":0.0000000032,\"112\":0.0,\"113\":0.000000002,\"114\":0.000000004,\"115\":0.0000000016,\"116\":0.000000001,\"117\":0.0000000019,\"118\":0.0000000023,\"119\":0.0,\"120\":0.0000000035,\"121\":0.0,\"122\":0.0,\"123\":0.0,\"124\":0.0,\"125\":0.0,\"126\":0.0,\"127\":0.0},\"fraction_ls_350\":{\"0\":2.412,\"1\":0.8362,\"2\":1.003,\"3\":1.4553,\"4\":5.027,\"5\":2.6649,\"6\":0.6371,\"7\":1.1822,\"8\":1.0407,\"9\":0.951,\"10\":1.5152,\"11\":2.1606,\"12\":13.8037,\"13\":2.8829,\"14\":2.4048,\"15\":3.0935,\"16\":2.819,\"17\":3.0328,\"18\":3.639,\"19\":3.9615,\"20\":0.5275,\"21\":1.1621,\"22\":1.1789,\"23\":1.3234,\"24\":1.4276,\"25\":1.3489,\"26\":2.1129,\"27\":1.3833,\"28\":1.6328,\"29\":5.9211,\"30\":52.9412,\"31\":44.2441,\"32\":27.8391,\"33\":26.1872,\"34\":25.6881,\"35\":12.7946,\"36\":4.1317,\"37\":39.4238,\"38\":32.5967,\"39\":35.5958,\"40\":26.9285,\"41\":24.3268,\"42\":15.4096,\"43\":6.947,\"44\":41.6136,\"45\":41.3055,\"46\":26.4506,\"47\":31.3549,\"48\":16.14,\"49\":5.0366,\"50\":30.9115,\"51\":33.541,\"52\":53.6614,\"53\":26.044,\"54\":5.3756,\"55\":7.4916,\"56\":39.8551,\"57\":31.7692,\"58\":23.3558,\"59\":20.7529,\"60\":13.6531,\"61\":12.7953,\"62\":9.1909,\"63\":6.7083,\"64\":43.8623,\"65\":27.2313,\"66\":17.5855,\"67\":14.2301,\"68\":12.8159,\"69\":9.1648,\"70\":46.3068,\"71\":28.6389,\"72\":14.2327,\"73\":13.5448,\"74\":11.7681,\"75\":7.2178,\"76\":23.1142,\"77\":16.6198,\"78\":11.9425,\"79\":10.0924,\"80\":7.9253,\"81\":38.5989,\"82\":16.4025,\"83\":40.5013,\"84\":48.1894,\"85\":30.7288,\"86\":30.5417,\"87\":39.5561,\"88\":50.2844,\"89\":1.0228,\"90\":1.6981,\"91\":1.6736,\"92\":1.56,\"93\":1.0273,\"94\":0.9409,\"95\":0.9456,\"96\":1.4679,\"97\":1.0305,\"98\":1.8778,\"99\":1.3505,\"100\":1.0687,\"101\":11.3665,\"102\":3.2993,\"103\":2.7324,\"104\":1.0458,\"105\":1.4911,\"106\":1.4587,\"107\":1.2556,\"108\":0.3693,\"109\":1.0959,\"110\":1.2934,\"111\":1.0484,\"112\":0.5773,\"113\":1.0745,\"114\":1.5471,\"115\":0.677,\"116\":0.5022,\"117\":0.257,\"118\":1.1314,\"119\":1.5735,\"120\":2.1706,\"121\":0.4845,\"122\":2.1111,\"123\":1.8714,\"124\":0.7168,\"125\":1.2799,\"126\":1.3941,\"127\":1.5964}};\n",
       "        var bait_string = d3.map(df['Bait_string']).entries().map(x=>x.value),\n",
       "            target_string = d3.map(df['Target_string']).entries().map(x=>x.value),\n",
       "            dataAxisX = d3.map(df['fraction_ls_350']).entries().map(x=>x.value/100),\n",
       "            dataAxisY = d3.map(df['hic_interaction']).entries().map(x=>x.value);\n",
       "        // get color for data in different chromosomes\n",
       "        var chromID = [];\n",
       "        var region = [];\n",
       "        for (var i = 0; i < bait_string.length;  i++) {\n",
       "            region = bait_string[i].split(/[-:]/);\n",
       "            chromID.push(region[0]);\n",
       "        }\n",
       "        // create custom data\n",
       "        var annoData = []\n",
       "        for (var i = 0; i < bait_string.length; i++) {\n",
       "            annoData.push([bait_string[i], target_string[i]]);\n",
       "        }\n",
       "        // prepare nb\n",
       "        var out = d3.select(element.get(0)).append('div')\n",
       "            .attr(\"id\", \"out\")\n",
       "        var nbHub = nb.dispatch(\"update\",\"brush\")\n",
       "        nbHub.chanId('cnbChan01')\n",
       "        nbHub.connect(function(d){\n",
       "          d3.select(\"#out\").html = d.connection\n",
       "        })\n",
       "        // plot data\n",
       "        var data = [\n",
       "            {\n",
       "                type: 'scatter',\n",
       "                mode: 'markers',\n",
       "                x: dataAxisX,\n",
       "                y: dataAxisY,\n",
       "                customdata: annoData,\n",
       "                hovertemplate: '<i>X</i>: %{x:.2f}' +\n",
       "                               '<br><i>Y</i>: %{y:.4e}' +\n",
       "                               '<br><i>Bait</i>: %{customdata[0]}' +\n",
       "                               '<br><i>Target</i>: %{customdata[1]}',\n",
       "                marker: {\n",
       "                    size:9,\n",
       "                    opacity: 0.7,\n",
       "                    line: {\n",
       "                        color: 'rgba(0,0,0,0.5)',\n",
       "                        width: 2\n",
       "                    }\n",
       "                },\n",
       "                transforms: [{type: \"groupby\", groups: chromID}],\n",
       "                showlegend: true\n",
       "            }\n",
       "        ]\n",
       "        var layout = {\n",
       "            hovermode:'closest',\n",
       "            title: \"Comparison of Hi-C and Microscopy datasets\",\n",
       "            xaxis: {\n",
       "                tickformat: '%',\n",
       "                title: \"Percentage of distance under 350nm\",\n",
       "            },\n",
       "            yaxis: {\n",
       "                type: 'log',\n",
       "                tickformat: '.1e',\n",
       "                autorange: true\n",
       "            },\n",
       "            legend: {orientation: 'h', y: -0.3}\n",
       "        };\n",
       "        // plot function\n",
       "        Plotly.newPlot('myDiv', data, layout);\n",
       "        // handle click event\n",
       "        myPlot.on('plotly_click', function(data) {\n",
       "            var bait = data.points[0].customdata[0];\n",
       "            var target = data.points[0].customdata[1];\n",
       "            var regionBait = bait.split(/[-:]/);\n",
       "            var regionTarget = target.split(/[-:]/);\n",
       "            // first navigate to that region\n",
       "            var navStart = Math.min(regionBait[1], regionTarget[1]) - 1000000;\n",
       "            var navEnd = Math.max(regionBait[2], regionTarget[2]) + 1000000;\n",
       "            nbHub.call(\"update\", this, [{\n",
       "                chr: regionBait[0], \n",
       "                start: navStart, \n",
       "                end: navEnd}]);\n",
       "            nbHub.call(\"brush\", this, [\n",
       "                {\n",
       "                chr: regionBait[0],\n",
       "                start: regionBait[1],\n",
       "                end: regionBait[2]\n",
       "                },\n",
       "                {\n",
       "                chr: regionTarget[0],\n",
       "                start: regionTarget[1],\n",
       "                end: regionTarget[2]\n",
       "                }])\n",
       "        })\n",
       "    });\n",
       "})(element);\n"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Javascript(\"\"\"\n",
    "(function(element){\n",
    "    require(['d3', 'nb', 'Plotly'], function(d3, nb, Plotly) {\n",
    "        // init div for plot\n",
    "        d3.select(element[0]).append('div')\n",
    "            .attr('id', 'myDiv')\n",
    "        var myPlot = document.getElementById('myDiv');\n",
    "        // prepare data\n",
    "        var df = %s;\n",
    "        var bait_string = d3.map(df['Bait_string']).entries().map(x=>x.value),\n",
    "            target_string = d3.map(df['Target_string']).entries().map(x=>x.value),\n",
    "            dataAxisX = d3.map(df['fraction_ls_350']).entries().map(x=>x.value/100),\n",
    "            dataAxisY = d3.map(df['hic_interaction']).entries().map(x=>x.value);\n",
    "        // get color for data in different chromosomes\n",
    "        var chromID = [];\n",
    "        var region = [];\n",
    "        for (var i = 0; i < bait_string.length;  i++) {\n",
    "            region = bait_string[i].split(/[-:]/);\n",
    "            chromID.push(region[0]);\n",
    "        }\n",
    "        // create custom data\n",
    "        var annoData = []\n",
    "        for (var i = 0; i < bait_string.length; i++) {\n",
    "            annoData.push([bait_string[i], target_string[i]]);\n",
    "        }\n",
    "        // prepare nb\n",
    "        var out = d3.select(element.get(0)).append('div')\n",
    "            .attr(\"id\", \"out\")\n",
    "        var nbHub = nb.dispatch(\"update\",\"brush\")\n",
    "        nbHub.chanId('cnbChan01')\n",
    "        nbHub.connect(function(d){\n",
    "          d3.select(\"#out\").html = d.connection\n",
    "        })\n",
    "        // plot data\n",
    "        var data = [\n",
    "            {\n",
    "                type: 'scatter',\n",
    "                mode: 'markers',\n",
    "                x: dataAxisX,\n",
    "                y: dataAxisY,\n",
    "                customdata: annoData,\n",
    "                hovertemplate: '<i>X</i>: %%{x:.2f}' +\n",
    "                               '<br><i>Y</i>: %%{y:.4e}' +\n",
    "                               '<br><i>Bait</i>: %%{customdata[0]}' +\n",
    "                               '<br><i>Target</i>: %%{customdata[1]}',\n",
    "                marker: {\n",
    "                    size:9,\n",
    "                    opacity: 0.7,\n",
    "                    line: {\n",
    "                        color: 'rgba(0,0,0,0.5)',\n",
    "                        width: 2\n",
    "                    }\n",
    "                },\n",
    "                transforms: [{type: \"groupby\", groups: chromID}],\n",
    "                showlegend: true\n",
    "            }\n",
    "        ]\n",
    "        var layout = {\n",
    "            hovermode:'closest',\n",
    "            title: \"Comparison of Hi-C and Microscopy datasets\",\n",
    "            xaxis: {\n",
    "                tickformat: '%%',\n",
    "                title: \"Percentage of distance under 350nm\",\n",
    "            },\n",
    "            yaxis: {\n",
    "                type: 'log',\n",
    "                tickformat: '.1e',\n",
    "                autorange: true\n",
    "            },\n",
    "            legend: {orientation: 'h', y: -0.3}\n",
    "        };\n",
    "        // plot function\n",
    "        Plotly.newPlot('myDiv', data, layout);\n",
    "        // handle click event\n",
    "        myPlot.on('plotly_click', function(data) {\n",
    "            var bait = data.points[0].customdata[0];\n",
    "            var target = data.points[0].customdata[1];\n",
    "            var regionBait = bait.split(/[-:]/);\n",
    "            var regionTarget = target.split(/[-:]/);\n",
    "            // first navigate to that region\n",
    "            var navStart = Math.min(regionBait[1], regionTarget[1]) - 1000000;\n",
    "            var navEnd = Math.max(regionBait[2], regionTarget[2]) + 1000000;\n",
    "            nbHub.call(\"update\", this, [{\n",
    "                chr: regionBait[0], \n",
    "                start: navStart, \n",
    "                end: navEnd}]);\n",
    "            nbHub.call(\"brush\", this, [\n",
    "                {\n",
    "                chr: regionBait[0],\n",
    "                start: regionBait[1],\n",
    "                end: regionBait[2]\n",
    "                },\n",
    "                {\n",
    "                chr: regionTarget[0],\n",
    "                start: regionTarget[1],\n",
    "                end: regionTarget[2]\n",
    "                }])\n",
    "        })\n",
    "    });\n",
    "})(element);\n",
    "\"\"\" % (data.to_json(orient = 'columns')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results\n",
    "\n",
    "The scatterplot in the notebook demonstrates the positive relationship between the Hi-C contact frequency (Y-axis) and the fraction of pair of genomic loci within 350nm (X-axis). Those pairs with high Hi-C contact frequency are also more likely to be close to each other. However, we also noticed that the fraction on the X-axis is low indicating high variability in single cells. \n",
    "\n",
    "We then select a pair on chromosome 1 and click it. The Nucloeme Browser (right) and the HiGlass panel in the middle automatically navigate to that regions and two loci with probes are highlighted in these two panels. In the Nucleome Browser, the bottom right is a specific panel hosting the raw images generated in the Finn et al. paper. Each track is one experiment with a different combination of probes. We select an image from the second track in which two probes are tested (red bars above the image track). The OMERO.iviwer window will pop out and we can further explore this raw FISH image.\n",
    "\n",
    "![Animation](img/NB_Jupyter_Finn_et_al_opt_1.8x.gif \"Ainimation 1\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
